File: projects/group/streamlit_app/guardian/agents/context_agent.py
"""Context Agent — 72-hour rolling snapshot of call/SMS/chat/txn events.

1:1 port of ``app/lib/agents/context_agent.dart``. Each ingest builds a
snapshot (recent event count, seconds since last call/SMS, prior max
risk score) and forwards to the risk agent for scoring.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import TYPE_CHECKING

from guardian.data.event_log import EventLog
from guardian.scenarios.events import (
    CallEvent,
    ChatEvent,
    ScamEvent,
    SmsEvent,
    TransactionEvent,
)

if TYPE_CHECKING:  # pragma: no cover - break import cycle
    from guardian.agents.risk_agent import RiskAgent
    from guardian.llm.tools import TraceCallback


_WINDOW = timedelta(hours=72)
_SENTINEL_SECONDS = 1 << 30


@dataclass(frozen=True)
class ContextSnapshot:
    triggering_event: ScamEvent
    recent_events: list[ScamEvent]
    now: datetime
    has_recent_call: bool
    has_recent_sms: bool
    has_recent_chat: bool
    seconds_since_last_call: int
    seconds_since_last_sms: int
    prior_max_risk: float

    @property
    def recent_event_count(self) -> int:
        return len(self.recent_events)


class ContextAgent:
    def __init__(self, event_log: EventLog, risk_agent: RiskAgent) -> None:
        self._log = event_log
        self._risk = risk_agent
        self.last_snapshot: ContextSnapshot | None = None
        self.trace_callback: TraceCallback | None = None
        self.trace_callback_factory: Callable[[ScamEvent], TraceCallback] | None = None

    def ingest(
        self,
        event: ScamEvent,
        trace_callback: TraceCallback | None = None,
    ) -> ContextSnapshot:
        self._log.add(event)
        recent = list(self._log.within(_WINDOW, now=event.timestamp))
        prior_max_risk = max(
            (
                entry.risk_score or 0.0
                for entry in self._log.entries
                if entry.event.id != event.id and entry.risk_score is not None
            ),
            default=0.0,
        )
        snapshot = self._build_snapshot(event, recent, prior_max_risk)
        self.last_snapshot = snapshot
        callback = trace_callback or self.trace_callback
        if callback is None and self.trace_callback_factory is not None:
            callback = self.trace_callback_factory(event)
        self._risk.assess(snapshot, trace_callback=callback)
        return snapshot

    @staticmethod
    def _build_snapshot(
        trigger: ScamEvent,
        recent: list[ScamEvent],
        prior_max_risk: float,
    ) -> ContextSnapshot:
        last_call: CallEvent | None = None
        last_sms: SmsEvent | None = None
        has_chat = False
        for e in recent:
            if isinstance(e, CallEvent):
                if last_call is None or e.timestamp > last_call.timestamp:
                    last_call = e
            elif isinstance(e, SmsEvent):
                if last_sms is None or e.timestamp > last_sms.timestamp:
                    last_sms = e
            elif isinstance(e, ChatEvent):
                has_chat = True
            elif isinstance(e, TransactionEvent):
                pass
        now = trigger.timestamp
        return ContextSnapshot(
            triggering_event=trigger,
            recent_events=list(recent),
            now=now,
            has_recent_call=last_call is not None,
            has_recent_sms=last_sms is not None,
            has_recent_chat=has_chat,
            seconds_since_last_call=(
                _SENTINEL_SECONDS
                if last_call is None
                else int((now - last_call.timestamp).total_seconds())
            ),
            seconds_since_last_sms=(
                _SENTINEL_SECONDS
                if last_sms is None
                else int((now - last_sms.timestamp).total_seconds())
            ),
            prior_max_risk=prior_max_risk,
        )

File: projects/group/legacy-flutter/app/lib/agents/context_agent.dart
import 'dart:developer' as dev;

import 'package:meta/meta.dart';
import 'package:riverpod/riverpod.dart';

import '../data/event_log.dart';
import '../scenarios/events.dart';
import 'risk_agent.dart';

@immutable
class ContextSnapshot {
  const ContextSnapshot({
    required this.triggeringEvent,
    required this.recentEvents,
    required this.now,
    required this.hasRecentCall,
    required this.hasRecentSms,
    required this.hasRecentChat,
    required this.secondsSinceLastCall,
    required this.secondsSinceLastSms,
    required this.priorMaxRisk,
  });

  final ScamEvent triggeringEvent;
  final List<ScamEvent> recentEvents;
  final DateTime now;
  final bool hasRecentCall;
  final bool hasRecentSms;
  final bool hasRecentChat;
  final int secondsSinceLastCall;
  final int secondsSinceLastSms;
  final double priorMaxRisk;

  int get recentEventCount => recentEvents.length;
}

class ContextAgent extends Notifier<ContextSnapshot?> {
  static const Duration _window = Duration(hours: 72);

  @override
  ContextSnapshot? build() => null;

  Future<void> ingest(ScamEvent event) async {
    final logNotifier = ref.read(eventLogProvider.notifier);
    logNotifier.add(event);
    final entries = ref.read(eventLogProvider);
    final recent = logNotifier.within(_window, now: event.timestamp).toList();
    final priorMaxRisk = entries
        .where((e) => e.event.id != event.id && e.riskScore != null)
        .map((e) => e.riskScore!)
        .fold<double>(0, (a, b) => a > b ? a : b);
    final snapshot = _buildSnapshot(event, recent, priorMaxRisk);
    state = snapshot;
    dev.log(
      '[context] ingested ${event.kind.name} — ${recent.length} event(s), '
      'prior max risk ${priorMaxRisk.toStringAsFixed(2)}',
      name: 'context',
    );
    await ref.read(riskAgentProvider.notifier).assess(snapshot);
  }

  ContextSnapshot _buildSnapshot(
    ScamEvent trigger,
    List<ScamEvent> recent,
    double priorMaxRisk,
  ) {
    CallEvent? lastCall;
    SmsEvent? lastSms;
    bool hasChat = false;
    for (final e in recent) {
      switch (e) {
        case CallEvent():
          if (lastCall == null || e.timestamp.isAfter(lastCall.timestamp)) {
            lastCall = e;
          }
        case SmsEvent():
          if (lastSms == null || e.timestamp.isAfter(lastSms.timestamp)) {
            lastSms = e;
          }
        case ChatEvent():
          hasChat = true;
        case TransactionEvent():
          break;
      }
    }
    final now = trigger.timestamp;
    return ContextSnapshot(
      triggeringEvent: trigger,
      recentEvents: recent,
      now: now,
      hasRecentCall: lastCall != null,
      hasRecentSms: lastSms != null,
      hasRecentChat: hasChat,
      secondsSinceLastCall: lastCall == null
          ? 1 << 30
          : now.difference(lastCall.timestamp).inSeconds,
      secondsSinceLastSms: lastSms == null
          ? 1 << 30
          : now.difference(lastSms.timestamp).inSeconds,
      priorMaxRisk: priorMaxRisk,
    );
  }
}

final contextAgentProvider =
    NotifierProvider<ContextAgent, ContextSnapshot?>(ContextAgent.new);
